### ISSUE TITLE
Debian Stretch build error with BIGNUM

### ISSUE THREAD

**Issue by srh** [rethinkdb/rethinkdb#6406]
I'm getting this build error on Debian Stretch, using `./configure --allow-fetch` in my sam/jsengine branch which uses Duktape to avoid GCC 6 problems.

It looks like the BIGNUM struct layout and size are supposed to be internal.

I think we should fix this in v2.3.x, v2.4.x, and next.

```
    [48/437] CC build/release/obj/rdb_protocol/geo/s2/s1interval.o
In file included from src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:3:0:
./src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.h:502:10: error: field ‘bn_’ has incomplete type ‘BIGNUM {aka bignum_st}’
   BIGNUM bn_;
          ^~~
In file included from /usr/include/openssl/bn.h:32:0,
                 from ./src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.h:103,
                 from src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:3:
/usr/include/openssl/ossl_typ.h:80:16: note: forward declaration of ‘BIGNUM {aka struct bignum_st}’
 typedef struct bignum_st BIGNUM;
                ^~~~~~~~~
In file included from src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:3:0:
./src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.h: In constructor ‘geo::ExactFloat::ExactFloat()’:
./src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.h:562:15: error: ‘BN_init’ was not declared in this scope
   BN_init(&bn_);
               ^
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc: In function ‘int geo::BN_ext_count_low_zero_bits(const BIGNUM*)’:
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:93:25: error: invalid use of incomplete type ‘const BIGNUM {aka const struct bignum_st}’
   for (int i = 0; i < bn->top; ++i) {
                         ^~
In file included from /usr/include/openssl/bn.h:32:0,
                 from ./src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.h:103,
                 from src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:3:
/usr/include/openssl/ossl_typ.h:80:16: note: forward declaration of ‘BIGNUM {aka struct bignum_st}’
 typedef struct bignum_st BIGNUM;
                ^~~~~~~~~
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:94:20: error: invalid use of incomplete type ‘const BIGNUM {aka const struct bignum_st}’
     BN_ULONG w = bn->d[i];
                    ^~
In file included from /usr/include/openssl/bn.h:32:0,
                 from ./src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.h:103,
                 from src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:3:
/usr/include/openssl/ossl_typ.h:80:16: note: forward declaration of ‘BIGNUM {aka struct bignum_st}’
 typedef struct bignum_st BIGNUM;
                ^~~~~~~~~
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc: In constructor ‘geo::ExactFloat::ExactFloat(double)’:
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:108:15: error: ‘BN_init’ was not declared in this scope
   BN_init(&bn_);
               ^
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc: In constructor ‘geo::ExactFloat::ExactFloat(int)’:
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:131:15: error: ‘BN_init’ was not declared in this scope
   BN_init(&bn_);
               ^
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc: In copy constructor ‘geo::ExactFloat::ExactFloat(const geo::ExactFloat&)’:
src/rdb_protocol/geo/s2/util/math/exactfloat/exactfloat.cc:143:15: error: ‘BN_init’ was not declared in this scope
   BN_init(&bn_);
```

**Comment by srh** [comment]
As a further complication, it looks like we try to access the members "top" and "d" of the BIGNUM struct.

**Comment by AtnNn** [comment]
From the openssl documentation:

> BN_init() was removed in OpenSSL 1.1.0; use BN_new() instead.

**Comment by srh** [comment]
A fix is in #6407.

**Comment by Extarys** [comment]
I see thx
